### PR TITLE
Pulling all storage-related methods into storage.js

### PR DESCRIPTION
**Pull request by redsquirrel**
Responding to #64

### CODE REVIEW

File: src/app/storage.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    var obj = {};
+    var done = false;
+    var count = 0
+    var dont = 0;
+
+    function check (key) {
+      chrome.storage.sync.get(key, function (resp) {
+        console.log('comparing to cloud', key, resp);
+        if (typeof resp[key] !== 'undefined' && obj[key] !== resp[key] && confirm('Overwrite "' + fileNameFromKey(key) + '"? Click Ok to overwrite local file with file from cloud. Cancel will push your local file to the cloud.')) {
```
Comment (axic) [inline src/app/storage.js]: The bug fixed by https://github.com/ethereum/browser-solidity/pull/68/commits/0b44735763adfbb0ea84a1c5642ccff6d5c68804 still applies here.

File: src/app/storage.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if (this.exists(filename) && this.get(filename) !== content) {
+      var count = '';
+      var otherKey = filename + count;
+      while (this.exists(filename + count)) count = count - 1;
+      this.rename(filename, filename + count);
+    }
+    this.set(filename, content);
+  };
+
+  this.sync = function () {
```
Comment (axic) [inline src/app/storage.js]: This is unrelated to this PR: is this `sync()` really about syncing with the "cloud"? Shouldn't it be named appropriately? Perhaps if its is Chrome only: `syncChromeCloud()`?

File: src/app/storage.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    var content = this.get(originalName);
+    this.set(newName, content);
+    this.remove(originalName);
+  };
+
+  this.remove = function (name) {
+    window.localStorage.removeItem(name);
+  };
+
+  this.setEditorSize = function (size) {
```
Comment (redsquirrel) [inline src/app/storage.js]: This encapsulates the storage/retrieval of the editor size from localStorage. In the future, we could create wrapper methods in editor.js if that makes more sense.

File: src/app/storage.js (hunk 2 above)
Comment (redsquirrel) [inline src/app/storage.js]: I haven't dug into what this is actually doing. I've just been trying to preserve behavior. I agree that its intent should be clarified.

File: src/app/storage.js (hunk 1 above)
Comment (redsquirrel) [inline src/app/storage.js]: Good point. Fixing...

File: src/app/storage.js (hunk 3 above)
Comment (axic) [inline src/app/storage.js]: `storage.get()` and `storage.set()` are public so it could be safely called from within the editor?

What it is used for? I don't see any references - though github diffs lately are weird and don't show everything.

Comment (redsquirrel) [inline src/app/storage.js]: I am encapsulating the `EDITOR_SIZE_CACHE_KEY` constant in storage.js, which is where it belongs since it's only used for localStorage.

`storage.setEditorSize` and `storage.getEditorSize` are both extracted (and now called) from app.js.

**Comment by chriseth** [comment]
Please rebas.

**Comment by redsquirrel** [comment]
Rebased

**Comment by chriseth** [comment]
Please rebase.

**Comment by redsquirrel** [comment]
I will rebase this once #77 is merged because I believe they conflict.

**Comment by redsquirrel** [comment]
Rebased

File: src/app/storage.js [hunk 4]
```diff
@@ -2,7 +2,56 @@
...
+
+  this.get = function (key) {
+    return window.localStorage.getItem(key);
+  };
+
+  this.keys = function () {
+    return Object.keys(window.localStorage);
+  };
+
+  this.loadFile = function (filename, content) {
```
Comment (chriseth) [inline src/app/storage.js]: It seems that `loadfile` has to be called with an already "encoded" filename, so its parameter should rather be called `key` and actually the function name should not refer to the concept of a "file".

I think it would be better to not expose the plain `get` and `set` methods here and only give access to file names (where this class would handle the encoding via `utils.fileKey`). If you say that this would be a subsequent refactoring step, I'm fine with that.

File: src/app/storage.js [hunk 5]
```diff
@@ -2,7 +2,56 @@
...
+function Storage (updateFiles) {
+  var EDITOR_SIZE_CACHE_KEY = 'editor-size-cache';
+
+  this.rename = function (originalName, newName) {
+    var content = this.get(originalName);
+    this.set(newName, content);
+    this.remove(originalName);
+  };
+
+  this.remove = function (name) {
```
Comment (chriseth) [inline src/app/storage.js]: Why not use prototypes?

File: src/app.js [hunk 6]
```diff
@@ -25,16 +25,13 @@ window.addEventListener('message', function (ev) {
 }, false);
 
 var run = function () {
+  var storage = new Storage(updateFiles);
+
   function loadFiles (files) {
     for (var f in files) {
       var key = utils.fileKey(f);
```
Comment (chriseth) [inline src/app.js]: No need for these two local variables.

File: src/app/compiler.js [hunk 7]
```diff
@@ -210,10 +210,7 @@ function Compiler (editor, renderer, queryParams, handleGithubCall, outputField,
           continue;
         }
         if (editor.hasFile(m)) {
-          files[m] = window.localStorage[utils.fileKey(m)];
-          reloop = true;
-        } else if (m.startsWith('./') && editor.hasFile(m.slice(2))) {
```
Comment (chriseth) [inline src/app/compiler.js]: Good that this is not needed anymore - what was the exact reason?

File: src/app/editor.js [hunk 8]
```diff
@@ -147,16 +151,15 @@ function Editor (loadingFromGist) {
 
   function setupStuff (files) {
     var untitledCount = '';
-    if (!files.length || window.localStorage['sol-cache']) {
+    if (files.length === 0) {
       if (loadingFromGist) return;
       // Backwards-compatibility
```
Comment (chriseth) [inline src/app/editor.js]: Actually we might not need that anymore - nobody used the old version at the new location.

**Comment by chriseth** [comment]
All comments are minor, can be merged as is, but please answer the questions.

File: src/app/storage.js (hunk 4 above)
Comment (redsquirrel) [inline src/app/storage.js]: Will be cleaned up in subsequent refactoring.

File: src/app/storage.js (hunk 5 above)
Comment (redsquirrel) [inline src/app/storage.js]: No particular reason. This approach works, and it's the approach I've used in all of my recent refactorings. If you'd like to rework those to use prototypes, create an issue an I'll investigate.

File: src/app.js (hunk 6 above)
Comment (redsquirrel) [inline src/app.js]: Are you referring to something that changed in this Pull Request? Or are you referring to `f` and `key`?

File: src/app/compiler.js (hunk 7 above)
Comment (redsquirrel) [inline src/app/compiler.js]: I'm lumping this into this Pull Request because I realized this code is now dead thanks to c2047b3b9482411446303bb8882d9f4d19421a29, which strips off leading `./`.

File: src/app/editor.js (hunk 8 above)
Comment (redsquirrel) [inline src/app/editor.js]: Right, so I'm getting rid of all of the `sol-cache` stuff, which I assume is from an early version of browser-solidity. Is there more that we can remove?

File: src/app.js (hunk 6 above)
Comment (chriseth) [inline src/app.js]: Yes, `key` and `content`

File: src/app/editor.js (hunk 8 above)
Comment (chriseth) [inline src/app/editor.js]: I think that's the only one for now.

File: src/app.js (hunk 6 above)
Comment (redsquirrel) [inline src/app.js]: Ah, I see. OK. I've committed a change that inlines those variables.

Comment (redsquirrel) [inline src/app.js]: Oops, you already merged. 😄  The "subsequent refactoring" has begun!
